### PR TITLE
[SPARK-8610] [SQL] Separate Row and InternalRow (part 2)

### PR DESCRIPTION
**Pull request by davies**
Currently, we use GenericRow both for Row and InternalRow, which is confusing because it could contain Scala type also Catalyst types.

This PR changes to use GenericInternalRow for InternalRow (contains catalyst types), GenericRow for Row (contains Scala types).

Also fixes some incorrect use of InternalRow or Row.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 1]
```diff
@@ -19,14 +19,54 @@ package org.apache.spark.sql.catalyst
...
+  override def getTimestamp(i: Int): java.sql.Timestamp = throw new UnsupportedOperationException
+  override def getSeq[T](i: Int): Seq[T] = throw new UnsupportedOperationException
+  override def getList[T](i: Int): java.util.List[T] = throw new UnsupportedOperationException
+  override def getMap[K, V](i: Int): scala.collection.Map[K, V] =
+    throw new UnsupportedOperationException
+  override def getJavaMap[K, V](i: Int): java.util.Map[K, V] =
+    throw new UnsupportedOperationException
+  override def getStruct(i: Int): InternalRow = throw new UnsupportedOperationException
+  override def getAs[T](fieldName: String): T = throw new UnsupportedOperationException
+  override def getValuesMap[T](fieldNames: Seq[String]): Map[String, T] =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: should we add `final` to make sure all `InternalRow` implementations can not define these functions?

**Comment by sujkh85** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-8610] [SQL] Separate Row and InternalRow (part 2) (#7003)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala (hunk 1 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: will do

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
(That test failure is my fault; already hotfixed)

**Comment by cloud-fan** [comment]
LGTM.

For data source API, I understand we need an efficient `buildScan` that return `RDD[InternalRow]` as `RDD[Row]`. But I'm still feel we should make `Row` and `InternalRow` siblings so that we can ensure that we don't forget to do the row transformation at compile time. I'll look into it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer.scala [hunk 2]
```diff
@@ -43,6 +41,7 @@ private[sql] class SparkSqlSerializer(conf: SparkConf) extends KryoSerializer(co
     kryo.setRegistrationRequired(false)
     kryo.register(classOf[MutablePair[_, _]])
     kryo.register(classOf[org.apache.spark.sql.catalyst.expressions.GenericRow])
+    kryo.register(classOf[org.apache.spark.sql.catalyst.expressions.GenericInternalRow])
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer.scala]: Do we need `SparkSqlSerializer` to serialize `GenericRow`?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer.scala]: I haven't looked the details, just follow the old code.

**Comment by davies** [comment]
In order to reduce the conflicts, I'm merging this into master.

**Pull request closed without merging**